structured_prediction_baselines/modules/sampler/sg_spen.py

```diff
@@ -0,0 +1,320 @@
+import contextlib
+import warnings
+from typing import (
+    List,
+    Tuple,
+    Union,
+    Dict,
+    Any,
+    Optional,
+    Callable,
+    Generator,
+)
+
+import numpy as np
+import torch
+import itertools
+from allennlp.common.lazy import Lazy
+from allennlp.training.optimizers import Optimizer
+
+from structured_prediction_baselines.modules.loss import Loss
+from structured_prediction_baselines.modules.oracle_value_function import (
+    OracleValueFunction,
+)
+from structured_prediction_baselines.modules.sampler import Sampler
+from structured_prediction_baselines.modules.stopping_criteria import (
+    StoppingCriteria,
+    StopAfterNumberOfSteps,
+)
+from structured_prediction_baselines.modules.score_nn import ScoreNN
+from structured_prediction_baselines.modules.task_nn import (
+    TaskNN,
+    CostAugmentedLayer,
+)
+
+
+@Sampler.register("sg-spen", constructor="from_partial_objects")
+class SGSpenSampler(Sampler):
```

Comment:
I see a lot of code copy-pasted from inference_net sampler. Should we inherit from InferenceNet sampler instead of base Sampler to avoid copying?